### PR TITLE
Device kernel for vector read_distributed

### PR DESCRIPTION
**Pull request by MarcelKoch**
This PR adds device kernel for distributed vector read_distributed. It mostly uses thrust/stl algorithms.

For now, this does not contain working dpcpp kernels. We will add them at a later point.

### CODE REVIEW

File: cuda/distributed/vector_kernels.cu [hunk 1]
```diff
@@ -36,22 +36,27 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 
 
-template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
-void build_local(
-    std::shared_ptr<const DefaultExecutor> exec,
-    const device_matrix_data<ValueType, GlobalIndexType>& input,
-    const distributed::Partition<LocalIndexType, GlobalIndexType>* partition,
-    comm_index_type local_part,
-    matrix::Dense<ValueType>* local_mtx) GKO_NOT_IMPLEMENTED;
+#define GKO_THRUST_LAMBDA __device__
```
Comment (upsj) [inline cuda/distributed/vector_kernels.cu:53]: I think we can just use __device__ directly, which is also shorter

File: common/cuda_hip/distributed/vector_kernels.hpp.inc [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    // write values with local rows into the local matrix at the correct index
+    // this needs the following iterators:
+    // - local_row_it: (global_row, range_id) -> local row index
+    // - flat_idx_it: (local_row, col) -> flat index in local matrix values
+    //                                    array
+    // the flat_idx_it is used by the scatter_if as an index map for the values
+    auto map_to_local_row =
+        [range_bounds, range_starting_indices] GKO_THRUST_LAMBDA(
+            const auto& idx_range_id) -> LocalIndexType {
```
Comment (upsj) [inline common/cuda_hip/distributed/vector_kernels.hpp.inc:62]: I saw some problematic interaction with generic lambdas and captures before. Maybe to be safe, state the type explicitly?
```suggestion
            thrust::tuple<...> idx_range_id) -> LocalIndexType {
```

**Review by upsj — COMMENTED** [review]
Nice job!

File: cuda/distributed/vector_kernels.cu (hunk 1 above)
Comment (upsj) [reply cuda/distributed/vector_kernels.cu:53]: could you try replacing the generic lambdas by explicit ones and see if they still occur? I haven't had any issues related to this so far.

Comment (MarcelKoch) [reply cuda/distributed/vector_kernels.cu:53]: Even with the generic lambdas I get the same error message. Btw, I'm using rocm 4.3, if that makes a difference. If you are interested, this is the error message:
```
/opt/rocm-4.3.0/include/thrust/detail/type_traits.h:455:26: error: no type named 'type' in 'thrust::detail::result_of_adaptable_function<(lambda at /home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:61:9) (thrust::tuple<int, unsigned long>)>'
  typedef typename Then::type type;
          ~~~~~~~~~~~~~~~^~~~
/opt/rocm-4.3.0/include/thrust/iterator/detail/iterator_adaptor_base.h:48:5: note: in instantiation of template class 'thrust::detail::eval_if<true, thrust::detail::result_of_adaptable_function<(lambda at /home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:61:9) (thrust::tuple<int, unsigned long>)>, thrust::detail::identity_<thrust::use_default>>' requested here
  : thrust::detail::eval_if<
    ^
/opt/rocm-4.3.0/include/thrust/iterator/detail/transform_iterator.inl:38:38: note: in instantiation of template class 'thrust::detail::ia_dflt_help<thrust::use_default, thrust::detail::result_of_adaptable_function<(lambda at /home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:61:9) (thrust::tuple<int, unsigned long>)>>' requested here
    typedef typename thrust::detail::ia_dflt_help<
                                     ^
/opt/rocm-4.3.0/include/thrust/iterator/transform_iterator.h:191:22: note: in instantiation of template class 'thrust::detail::transform_iterator_base<(lambda at /home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:61:9), thrust::zip_iterator<thrust::tuple<const int *, unsigned long *>>, thrust::use_default, thrust::use_default>' requested here
    : public detail::transform_iterator_base<AdaptableUnaryFunction, Iterator, Reference, Value>::type
                     ^
/home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:68:25: note: in instantiation of template class 'thrust::transform_iterator<(lambda at /home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:61:9), thrust::zip_iterator<thrust::tuple<const int *, unsigned long *>>>' requested here
    auto local_row_it = thrust::make_transform_iterator(
                        ^
/home/mkoch/projects/worktrees/ginkgo/distributed-vector-kernels/common/cuda_hip/distributed/vector_kernels.hpp.inc:94:5: note: in instantiation of function template specialization 'gko::kernels::hip::distributed_vector::build_local<std::complex<double>, int, int>' requested here
    GKO_DECLARE_DISTRIBUTED_VECTOR_BUILD_LOCAL);
```

Comment (upsj) [reply cuda/distributed/vector_kernels.cu:53]: `__host__ __device__` everywhere then? :) I guess I just didn't use transform_iterator on ROCm before, contrary to what I thought

Comment (MarcelKoch) [reply cuda/distributed/vector_kernels.cu:53]: Yes, I will use that. I didn't use it before, because `__host__ __device__` doesn't work with cuda and generic lambdas...

**Comment by MarcelKoch** [comment]
format!

**Comment by MarcelKoch** [comment]
format!

File: common/cuda_hip/distributed/vector_kernels.hpp.inc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    // write values with local rows into the local matrix at the correct index
+    // this needs the following iterators:
+    // - local_row_it: (global_row, range_id) -> local row index
+    // - flat_idx_it: (local_row, col) -> flat index in local matrix values
+    //                                    array
+    // the flat_idx_it is used by the scatter_if as an index map for the values
+    auto map_to_local_row =
+        [range_bounds, range_starting_indices] __host__ __device__(
+            const thrust::tuple<GlobalIndexType, size_type>& idx_range_id)
+        -> LocalIndexType {
```
Comment (upsj) [inline common/cuda_hip/distributed/vector_kernels.hpp.inc:63]: the return type shouldn't be necessary, since you already static_cast to the type and add another LocalIndexType
```suggestion
            const thrust::tuple<GlobalIndexType, size_type>& idx_range_id) {
```

File: test/distributed/vector_kernels.cpp [hunk 4]
```diff
@@ -68,11 +71,17 @@ class Vector : public ::testing::Test {
...
     using mtx = gko::matrix::Dense<value_type>;
 
-    Vector()
-        : ref(gko::ReferenceExecutor::create()),
-          exec(gko::OmpExecutor::create()),
-          engine(42)
-    {}
+    Vector() : ref(gko::ReferenceExecutor::create()), engine(42) {}
+
+
```
Comment (upsj) [inline test/distributed/vector_kernels.cpp:76]: nit:
```suggestion
```

File: test/mpi/distributed/vector.cpp [hunk 5]
```diff
@@ -420,11 +418,7 @@ class VectorReductions : public ::testing::Test {
         y = gko::clone(exec, tmp_y);
     }
 
-    void SetUp() override
-    {
-        ASSERT_GT(comm.size(), 0);
-        init_executor(gko::ReferenceExecutor::create(), exec);
-    }
+    void SetUp() override { ASSERT_GT(comm.size(), 0); }
```
Comment (upsj) [inline test/mpi/distributed/vector.cpp:421]: What's the reason for these changes?

File: third_party/gtest/CMakeLists.txt [hunk 6]
```diff
@@ -3,7 +3,7 @@ include(FetchContent)
 FetchContent_Declare(
     googletest
     GIT_REPOSITORY https://github.com/google/googletest.git
-    GIT_TAG        release-1.11.0
+    GIT_TAG        main
```
Comment (upsj) [inline third_party/gtest/CMakeLists.txt:6]: I think it's better to tag a fixed version. Do you need a more recent version?

**Review by upsj — APPROVED** [review]
LGTM!

File: common/cuda_hip/distributed/vector_kernels.hpp.inc [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        partition->get_range_starting_indices();
+    const auto* part_ids = partition->get_part_ids();
+    const auto num_ranges = partition->get_num_ranges();
+
+    Array<size_type> range_id{exec, input.get_num_elems()};
+    thrust::upper_bound(thrust::device, range_bounds + 1,
+                        range_bounds + num_ranges + 1,
+                        input.get_const_row_idxs(),
+                        input.get_const_row_idxs() + input.get_num_elems(),
+                        range_id.get_data(), thrust::less<GlobalIndexType>());
```
Comment (pratikvn) [inline common/cuda_hip/distributed/vector_kernels.hpp.inc:52]: Will the difference in types of `range_id` and `GlobalIndexType` make a difference here ?

**Review by pratikvn — APPROVED** [review]
Nice work! LGTM!

File: test/mpi/distributed/vector.cpp (hunk 5 above)
Comment (MarcelKoch) [reply test/mpi/distributed/vector.cpp:421]: Because I would need to re-add the logger to exec here, but instead I just do it once in the constructor.

File: third_party/gtest/CMakeLists.txt (hunk 6 above)
Comment (MarcelKoch) [reply third_party/gtest/CMakeLists.txt:6]: My bad, that should not be part of the PR
